packages/next/build/webpack/loaders/next-middleware-ssr-loader/index.ts

```diff
@@ -22,6 +44,18 @@ export default function middlewareRSCLoader(this: any) {
     './pages/_app'
   )
 
+  const fs = this.fs.fileSystem
+  const hasProvidedAppPage = fs.existsSync(stringifiedAbsoluteAppPath)
+  const hasProvidedDocumentPage = fs.existsSync(stringifiedAbsoluteDocumentPath)
```

Comment:
Other parts look good to me! Only thing is that we shouldn't use sync APIs, maybe we have to change this loader to async and use `fs.exists`.